src/core/search/search_test.cc

```diff
@@ -0,0 +1,279 @@
+// See LICENSE for licensing terms.
+//
+
+#include "core/search/search.h"
+
+#include <absl/container/flat_hash_map.h>
+
+#include <algorithm>
+#include <random>
+
+#include "base/gtest.h"
+#include "base/logging.h"
+#include "core/search/base.h"
+#include "core/search/query_driver.h"
+
+namespace dfly {
+namespace search {
+
+using namespace std;
+
+struct MockedDocument : public DocumentAccessor {
+ public:
+  using Map = absl::flat_hash_map<std::string, std::string>;
+
+  MockedDocument() = default;
+  MockedDocument(Map map) : fields_{map} {
+  }
+  MockedDocument(std::string test_field) : fields_{{"field", test_field}} {
+  }
+
+  bool Check(DocumentAccessor::FieldConsumer f, string_view active_field) const override {
+    if (!active_field.empty()) {
+      auto it = fields_.find(string{active_field});
+      return f(it != fields_.end() ? it->second : "");
+    } else {
+      for (const auto& [k, v] : fields_) {
+        if (f(v))
+          return true;
+      }
+      return false;
+    }
+  }
+
+  string DebugFormat() {
+    string out = "{";
+    for (const auto& [field, value] : fields_)
+      absl::StrAppend(&out, field, "=", value, ",");
+    if (out.size() > 1)
+      out.pop_back();
+    out += "}";
+    return out;
+  }
+
+  void Set(Map hset) {
+    fields_ = hset;
+  }
+
+ private:
+  Map fields_{};
+};
+
+// Just a filler. TODO: Remove
+struct Schema {
+  enum FieldType { TEXT, NUMERIC };
+
+  absl::flat_hash_map<std::string, FieldType> fields;
+};
+
+class SearchParserTest : public ::testing::Test {
+ protected:
+  SearchParserTest() {
+    PrepareSchema();
+  }
+
+  ~SearchParserTest() {
+    EXPECT_EQ(entries_.size(), 0u) << "Missing check";
+  }
+
+  void PrepareSchema(Schema schema = {{{"field", Schema::TEXT}}}) {
+    schema_ = schema;
+  }
+
+  void PrepareQuery(string_view query) {
+    query_ = query;
+  }
+
+  template <typename... Args> void ExpectAll(Args... args) {
+    (entries_.emplace_back(args, true), ...);
+  }
+
+  template <typename... Args> void ExpectNone(Args... args) {
+    (entries_.emplace_back(args, false), ...);
+  }
+
+  bool Check() {
+    SearchAlgorithm search_algo{};
+    search_algo.Init(query_);
+
+    for (auto& [doc, expect_match] : entries_) {
+      bool doc_matched = search_algo.Check(&doc);
+
+      if (doc_matched != expect_match) {
+        error_ = "doc: \"" + doc.DebugFormat() + "\"" + " was expected" +
+                 (expect_match ? "" : " not") + " to match" + " query: \"" + query_ + "\"";
+        entries_.clear();
+        return false;
+      }
+    }
+
+    entries_.clear();
+    return true;
+  }
+
+  string_view GetError() const {
+    return error_;
+  }
+
+ private:
+  using DocEntry = pair<MockedDocument, bool /*should_match*/>;
+
+  Schema schema_;
+  vector<DocEntry> entries_;
+  string query_, error_;
+};
+
+TEST_F(SearchParserTest, MatchTerm) {
+  PrepareQuery("foo");
+
+  // Check basic cases
+  ExpectAll("foo", "foo bar", "more foo bar");
```

Comment:
the problem with that is that you won't know where it fails, what line.
I suggest to use traditional methods with matchers and macros

Reply:
See my comment, but I'll repeat it here.

You will see what test fails (including subtest by line). It will print enough info (what case was matched inversely). I don't believe it makes a difference in comparison with per-line failures. You get the same info.

When writing query-centric tests, **I can't assert per line, because I first want to insert all test cases into the index to simulate multi value search**. AFAIK matchers can't transfer context that easily (first execute and only then fail on a line). 

The other option is writing index-centric tests like in `search_family_test`. First, build the index, then assert query results per-line. I can do it, but I'm not fond of it, because:
* its easier to check functionality with query-centric tests. You know the execution flow, you know what tests fit bests
* all those tests need to be rewritten